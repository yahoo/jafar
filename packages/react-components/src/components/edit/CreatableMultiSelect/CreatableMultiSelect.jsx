/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import SelectInternal from 'react-select/creatable';
import { customStyles } from '../Select/Select';

/**
 * Represent creation of dynamic string array
 * 
 * Import <a target="_blank" href=
 "https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/edit/
 CreatableMultiSelect/CreatableMultiSelect.jsx">
 CreatableMultiSelect</a> from '@jafar/react-components/edit/CreatableMultiSelect'
 */
export default class CreatableMultiSelect extends React.Component {
  static propTypes = {
    value: PropTypes.arrayOf(PropTypes.string),
    state: PropTypes.shape({
      placeholder: PropTypes.string,
    }),
    disabled: PropTypes.bool,
    onValueChange: PropTypes.func.isRequired,
  };

  static defaultProps = {
    value: [],
    state: {
      placeholder: '',
    },
    disabled: false,
  };

  render() {
    const value = (this.props.value || []).map(item => { return { label: item, value: item }; });

    return (
      <SelectInternal
        isMulti={true}
        value={value} 
        onChange={this.onValueChange}
        placeholder={this.props.state.placeholder}
        isDisabled={this.props.disabled}
        styles={customStyles}
      />
    );
  }

  onValueChange = (selected) => {
    const newValue = !selected ? undefined : selected.map(item => item.value);
    this.props.onValueChange(newValue);
  }
}
