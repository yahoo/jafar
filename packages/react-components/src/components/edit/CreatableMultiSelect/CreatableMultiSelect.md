<h3>Examples</h3>

<h4>Usage in jafar form</h4>

```javascript
const Form = require('@jafar/react-form/Form').default;
const Field = require('@jafar/react-form/Field').default;

const model = {
  id: 'simple',
  fields: {
    keywords: {
      label: 'Keywords',
      path: 'keywords',
      component: {
        name: 'CreatableMultiSelect',
      },
      required: true,
      disableTerm: {
          name: 'disableMe',
      },
      excludeTerm: {
          name: 'excludeMe',
      },
    },
  },
  data: {
      keywords: ['dog', 'cat'], // remove to see initial value undefined, and required error
  }
};

const resources = {
  components: { 
    CreatableMultiSelect: {
      renderer: CreatableMultiSelect,
    }, 
  },
  terms: {
      disableMe: { func: () => false }, // convert to true to see the field disabled
      excludeMe: { func: () => false }, // convert to true to see the field excluded
  },
};

<Form model={model} resources={resources}>
    <Field id='keywords' />
</Form>
```

<h4>Simple</h4>

```javascript
initialState = { 
  value: ['foo', 'bar'],
};

<CreatableMultiSelect
value={state.value}
state={state.state}
onValueChange={(value) => {
  setState({ value });
}}
/>
```