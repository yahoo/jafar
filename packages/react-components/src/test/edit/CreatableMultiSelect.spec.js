/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { mount, shallow } from 'enzyme';
import CreatableMultiSelect from '../../components/edit/CreatableMultiSelect/index';

describe('<CreatableMultiSelect />', () => {
  let component;
  let onValueChangeSpy;
  let onStateChangeSpy;
  let state;
  let value;
  let disabled;
  let required;

  beforeEach(() => {
    onValueChangeSpy = jest.fn();
    onStateChangeSpy = jest.fn();
    value = ['ocean'];
    disabled = false;
    required = false;
    state = { };
  });

  it('Should render provided data', () => {
    component = shallow(
      getComponent(value, state, disabled, required, onValueChangeSpy, onStateChangeSpy)
    );
    expect(component).toMatchSnapshot();
  });

  it('Should render provided data - disabled', () => {
    disabled = true;
    component = shallow(
      getComponent(value, state, disabled, required, onValueChangeSpy, onStateChangeSpy)
    );
    expect(component).toMatchSnapshot();
  });

  it('Should render provided data - required', () => {
    required = true;
    component = shallow(
      getComponent(value, state, disabled, required, onValueChangeSpy, onStateChangeSpy)
    );
    expect(component).toMatchSnapshot();
  });

  it('onValueChange - new value defined - ok', () => {
    component = mount(
      getComponent(value, state, disabled, required, onValueChangeSpy, onStateChangeSpy)
    );
    const controller = component.instance();
    const newValue = [{ value: 'purple', label: 'purple' }];
    controller.onValueChange(newValue);
    expect(onValueChangeSpy).toHaveBeenCalledWith(['purple']);
  });

  it('onValueChange - new value undefined - ok', () => {
    component = mount(
      getComponent(value, state, disabled, required, onValueChangeSpy, onStateChangeSpy)
    );
    const controller = component.instance();
    const newValue = undefined;
    controller.onValueChange(newValue);
    expect(onValueChangeSpy).toHaveBeenCalledWith(undefined);
  });

  function getComponent(value, state, disabled, required, onValueChange, onStateChange) {
    return (
      <CreatableMultiSelect
        value={value}
        state={state}
        disabled={disabled}
        required={required}
        onValueChange={onValueChange}
        onStateChange={onStateChange}
      />
    );
  }
});
