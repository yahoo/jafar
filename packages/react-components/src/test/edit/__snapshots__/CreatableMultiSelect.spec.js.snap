// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreatableMultiSelect /> Should render provided data - disabled 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  isDisabled={true}
  isMulti={true}
  onChange={[Function]}
  styles={
    Object {
      "clearIndicator": [Function],
      "control": [Function],
      "dropdownIndicator": [Function],
      "menu": [Function],
      "menuList": [Function],
      "singleValue": [Function],
      "valueContainer": [Function],
    }
  }
  value={
    Array [
      Object {
        "label": "ocean",
        "value": "ocean",
      },
    ]
  }
/>
`;

exports[`<CreatableMultiSelect /> Should render provided data - required 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  isDisabled={false}
  isMulti={true}
  onChange={[Function]}
  styles={
    Object {
      "clearIndicator": [Function],
      "control": [Function],
      "dropdownIndicator": [Function],
      "menu": [Function],
      "menuList": [Function],
      "singleValue": [Function],
      "valueContainer": [Function],
    }
  }
  value={
    Array [
      Object {
        "label": "ocean",
        "value": "ocean",
      },
    ]
  }
/>
`;

exports[`<CreatableMultiSelect /> Should render provided data 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  isDisabled={false}
  isMulti={true}
  onChange={[Function]}
  styles={
    Object {
      "clearIndicator": [Function],
      "control": [Function],
      "dropdownIndicator": [Function],
      "menu": [Function],
      "menuList": [Function],
      "singleValue": [Function],
      "valueContainer": [Function],
    }
  }
  value={
    Array [
      Object {
        "label": "ocean",
        "value": "ocean",
      },
    ]
  }
/>
`;
